### PR TITLE
Hardcode known wrapper checksums to avoid network requests

### LINKED ISSUE #161: Hardcode list of known checksums to avoid network requests in most cases
This goes in the same direction as #96; was also suggested in https://github.com/gradle/wrapper-validation-action/issues/57#issuecomment-1135512701

What do you think about hardcoding the list of all known Gradle wrapper checksums into this action? Then during validation, the action first checks if the checksum is in that hardcoded list, or is specified in `allow-checksums`. And only if both checks fail it sends network requests to fetch the checksums.

This would hopefully have the following advantages:
- Action is more efficient and reduces network traffic for Gradle site
- Fewer spurious action failures due to network issues

It would not be an issue if the hardcoded list of checksums becomes outdated (possibly by months or even years); it probably still provides an advantage for the majority of the users, assuming they use older wrapper versions.

### Alternative
Tell users to use the `allow-checksums` input. However, that risks reducing security because it might then become a usual action to update the `allow-checksums` input when updating the wrapper, and maintainers might then not properly verify that the `allow-checksums` is actually one of the officially listed ones. This allows a malicious user to simply specify the checksum of their malicious wrapper, and claim it is the official checksum, hoping the maintainer does not verify it.

### PR DESCRIPTION
**Pull request by Marcono1234**
Fixes #161

If a checksum is not found in the hardcoded list, the action falls back to fetching the checksums from the Gradle API, as before.

For now there is no logic for automatically updating the list of known checksums; that has to be done manually. But maybe it could be automated in some way in the future.

Here is my somewhat hacky (but hopefully bug-free) Java code which I used to generate the list of checksums:
<details>

<summary>Checksums list creator</summary>

```java
import java.io.IOException;
import java.lang.reflect.Type;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import com.google.gson.Gson;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.annotations.JsonAdapter;
import com.google.gson.reflect.TypeToken;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class GradleChecksumsFetcher {
    private static final OkHttpClient client = new OkHttpClient();

    private static String fetch(String url) throws IOException {
      Request request = new Request.Builder()
          .url(url)
          .build();

      try (Response response = client.newCall(request).execute()) {
        return response.body().string();
      }
    }

    static class VersionData {
        String version;
        String wrapperChecksumUrl;
        @JsonAdapter(BuildTimeAdapter.class)
        Instant buildTime;

        boolean snapshot;
        boolean nightly;
        boolean releaseNightly;

        private static class BuildTimeAdapter implements JsonDeserializer<Instant> {
            private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyyMMddHHmmssxx", Locale.ROOT);

            @Override
            public Instant deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                    throws JsonParseException {
                return FORMATTER.parse(json.getAsString(), Instant::from);
            }
        }
    }

    private static List<String> parseVersion(String v) {
        int i = v.indexOf('-');
        if (i != -1) {
            v = v.substring(0, i);
        }

        return List.of(v.split("\\."));
    }

    private static int compareVersions(String a, String b) {
        List<String> aParts = parseVersion(a);
        List<String> bParts = parseVersion(b);

        for (int i = 0; i < Math.min(aParts.size(), bParts.size()); i++) {
            int aPart = Integer.parseInt(aParts.get(i));
            int bPart = Integer.parseInt(bParts.get(i));

            if (aPart < bPart) {
                return -1;
            }
            if (aPart > bPart) {
                return 1;
            }
        }
        return Integer.compare(aParts.size(), bParts.size());
    }

    public static void main(String[] args) throws Exception {
        Gson gson = new Gson();
        List<VersionData> versions = gson.fromJson(fetch("https://services.gradle.org/versions/all"), new TypeToken<>() {});
        // First sort by version number, then by build time
        // That should (in most cases) achieve version ranges where all hashes only belong to one range
        Comparator<VersionData> comparator = (v1, v2) -> compareVersions(v1.version, v2.version);
        versions.sort(comparator.thenComparing(Comparator.comparing(v -> v.buildTime)));

        String firstVersionName = null;
        String lastVersionName = null;
        String lastHash = null;

        Set<String> seenHashes = new HashSet<>();

        for (VersionData version : versions) {
            if (version.wrapperChecksumUrl == null || version.snapshot || version.nightly || version.releaseNightly) {
                continue;
            }
            String versionName = version.version;
            String hash = fetch(version.wrapperChecksumUrl);

            if (lastHash == null) {
                firstVersionName = versionName;
                lastVersionName = versionName;
                lastHash = hash;
            } else {
                if (hash.equals(lastHash)) {
                    lastVersionName = versionName;
                } else {
                    if (firstVersionName.equals(lastVersionName)) {
                        System.out.println("// " + firstVersionName);
                    } else {
                        System.out.println("// " + firstVersionName + " - " + lastVersionName);
                    }
                    System.out.println("\"" + lastHash + "\",");

                    firstVersionName = versionName;
                    lastVersionName = versionName;
                    lastHash = hash;

                    // This acts mainly as assertion that the version sorting logic is correct
                    // There seems to be one case though where the version range `8.0-milestone-1 - 8.0-milestone-3`
                    // is using the same hash as a previous version range
                    if (!seenHashes.add(hash)) {
                        System.out.println("// WARNING: Duplicate hash: " + hash + "; for version " + versionName);
                    }
                }
            }
        }

        System.out.println("// " + firstVersionName + " - " + lastVersionName);
        System.out.println("\"" + lastHash + "\",");
    }

}
```

</details>

:warning: I am not that familiar with TypeScript, so any feedback is appreciated!

### CODE REVIEW

File: src/checksums.ts [hunk 1]
```diff
@@ -6,9 +6,165 @@ const httpc = new httpm.HttpClient(
   {allowRetries: true, maxRetries: 3}
 )
 
+/** Known checksums from previously published Wrapper versions */
+export const KNOWN_VALID_CHECKSUMS = new Set([
```
Comment (JLLeitschuh) [inline src/checksums.ts:10]: Is there a way to pull this from a resource file instead of hard-coding it into the source code.

Ideally, there would be a resource file, and there would also be a script that CI could automatically run to update this file and would auto-push then auto-release when there was an update

**Review by JLLeitschuh — APPROVED** [review]
In general, this looks like it's headed in the right direction!

I think the biggest ask is tooling/automation to keep the list up-to-date more easily, and ideally, with a GitHub action

File: src/checksums.ts (hunk 1 above)
Comment (bigdaz) [reply src/checksums.ts:10]: Yep, that would be much better. If the file was formatted as JSON, it could be trivially loaded into an object: that way we could include the Gradle version as well as the checksum in the metadata.

@Marcono1234 you can [see here](https://github.com/gradle/actions/blob/main/sources/src/cache-base.ts#L236-L241) an example of loading a resource file in Typescript. It may not be the best way, but it works!
Then, you can use `JSON.parse(fileContents)` and use it as a typed Javascript object. Here [are](https://github.com/gradle/actions/blob/aff52e5be96935327d77c5529075184377dc4371/sources/src/build-results.ts#L25) [some](https://github.com/gradle/actions/blob/aff52e5be96935327d77c5529075184377dc4371/sources/src/provision.ts#L168) [examples](https://github.com/gradle/actions/blob/aff52e5be96935327d77c5529075184377dc4371/sources/src/cache-extract-entries.ts#L281).

**Review by bigdaz — CHANGES_REQUESTED** [review]
This looks pretty good to me. If we can extract the checksums list into a resource file, then I'm OK if this needs to be maintained by hand (for now). New Gradle versions will continue to require a network call until the wrapper action is updated.

As @JLLeitschuh mentioned, it would be ideal if we had a job that ran in this repository to update the versions file when a new Gradle version is released. But we'll still need to release a new action version each time.

An additional optimization would be to store any "discovered" checksums in the GitHub Actions cache. We read the set of known versions from the resource, add any discovered versions, and write the JSON file directly to the cache.

**Comment by bigdaz** [comment]
@Marcono1234 The changes to `dist/index.js` make the PR very difficult to merge. Can you please update the commit to remove that change?

**Comment by Marcono1234** [comment]
Thanks for the feedback! I will try to adjust this in the next days.

Would it make sense though to use a plaintext / custom text format where for example each line represents a checksum, except if it is blank or starts with `#`? For example something like:
```
# 1.0
87e50531ca7aab675f5bb65755ef78328afd64cf0877e37ad876047a8a014055
# 1.1
22c56a9780daeee00e5bf31621f991b68e73eff6fe8afca628a1fe2c50c6038e
# 1.2
5c91fa893665f3051eae14578fac2df14e737423387e75ffbeccd35f335a3d8b
...
```
JSON has the disadvantage that it doesn't support comments, and just having a large list of checksums without any indication to which versions they belong might make reviewing it difficult.

However, with #145 maybe it would make sense to use JSON instead of a custom text format, but then include the version numbers. For example:
```json
{
  "87e50531ca7aab675f5bb65755ef78328afd64cf0877e37ad876047a8a014055": [
    "1.0"
  ],
  "22c56a9780daeee00e5bf31621f991b68e73eff6fe8afca628a1fe2c50c6038e": [
    "1.1",
  ]
}
```

**Comment by bigdaz** [comment]
> However, with https://github.com/gradle/wrapper-validation-action/pull/145 maybe it would make sense to use JSON instead of a custom text format, but then include the version numbers.

Yes that's what I meant. Something like: 
```
[
 { "version": "1.0", "checksum": "....." },
 { "version": "1.1", "checksum": "....." },
  ... etc ..
]
```
That way you could use `JSON.parse` to read this directly into an array of Typescript object with a `version` and `checksum` attribute. Take a look at the links I posted for examples.

Once you have this array, you can get the list of checksums on their own using `arrayOfObjects.map { it.checksum }`.

**Comment by Marcono1234** [comment]
I have performed the following changes now:
- Included checksums as JSON file
The JSON file is imported as module instead of using `path.resolve(__dirname, ...)`. This seems more reliable to me because depending on whether unit tests or the action (from `index.js`) is run `__dirname` differs (I assume), and therefore both `src/...` and `dist/...` would need to have the same nesting level for this to work.
- Changed known `KNOWN_VALID_CHECKSUMS` to be a `Map<string, Set<string>>`, mapping from checksum to set of version names. This is for compatibility with [#145](https://github.com/gradle/wrapper-validation-action/pull/145); could then for example calculate the checksum for the JAR, and check whether the expected version is in the set for that checksum.
- Added GitHub workflow for updating checksums file and creating a pull request
This runs every week but can also be run manually. Here is an example how the created pull request looks like: https://github.com/Marcono1234/wrapper-validation-action/pull/6.

I hope that is ok like this. Feedback is appreciated!

File: .github/workflows/update-checksums-file.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          npm install typed-rest-client@1.8.11 --no-save
+
+      - name: Update checksums file
+        run: node ./.github/workflows/update-checksums-file.js
+
+      # If there are no changes, this action will not create a pull request
+      - name: Create or update pull request
+        uses: peter-evans/create-pull-request@v6
+        with:
+          branch: wrapper-checksums-update
```
Comment (Marcono1234) [inline .github/workflows/update-checksums-file.yml:36]: Maybe the branch name should use a prefix such as `bot/` to prevent accidentally manually editing this branch, which might then be overwritten by this workflow.

What do you think?

**Comment by bigdaz** [comment]
Thanks @Marcono1234 . I'm not going to have capacity to take this further until Feb 13 at the earliest. But it's on my list!

File: .github/workflows/update-checksums-file.js
Comment (JLLeitschuh) [inline .github/workflows/update-checksums-file.js:1]: Awesome. Was thinking about writing this myself and wasn't looking forward to it. Solid bit of work here!

Javascript is definitely not my strongest area, ngl. 😆

File: .github/workflows/update-checksums-file.yml (hunk 2 above)
Comment (JLLeitschuh) [reply .github/workflows/update-checksums-file.yml:36]: Sounds completely reasonable to prefix it with `bot/`

File: tsconfig.json [hunk 3]
```diff
@@ -6,7 +6,8 @@
     "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
     "strict": true,                           /* Enable all strict type-checking options. */
     "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
-    "esModuleInterop": true                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "esModuleInterop": true,                  /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "resolveJsonModule": true,                /* Enable importing JSON files as module; used for importing wrapper checksums JSON */
```
Comment (JLLeitschuh) [inline tsconfig.json:10]: This was what I was missing when I took an early crack at this! 🎉

File: src/checksums.ts [hunk 4]
```diff
@@ -1,14 +1,40 @@
...
+
+  return versionsMap
+}
+
+/**
+ * Known checksums from previously published Wrapper versions.
+ *
+ * Maps from the checksum to the names of the Gradle versions whose wrapper has this checksum.
+ */
+export const KNOWN_VALID_CHECKSUMS = getKnownValidChecksums()
```
Comment (JLLeitschuh) [inline src/checksums.ts:33]: Can you add a unit test that verifies that this is never empty, and always contains some reasonable versions as a spot-check that this logic never breaks?

**Comment by bigdaz** [comment]
This has been merged manually. I need to find a better process for merging external PRs.

**Comment by bigdaz** [comment]
> 'd like to include this in the first RC of v2.0.0, since that mitigates some of the risk of the change. (People have to explicitly update to get the new version).

Actually, I forgot that I already released `v2.0.0-rc.1` 🤦🏼 .
This change will need to wait for a `v2.1.0` release.

**Comment by JLLeitschuh** [comment]
@Marcono1234 seriously, thank you so much for working this problem out. Really truly. You can see by the number of issues that @bigdaz closed how much pain our network connection logic has caused over the years.

You've just significantly increased the stability and usability of this action for all of the users of it.

Truly, thank you so much for your contribution. I'm incredibly appreciative you took the time to contribute it.

**Comment by Marcono1234** [comment]
Thanks for the kind words and the feedback!

I have addressed the feedback (adding tests, and `bot/` branch prefix) in #178.

**Comment by bigdaz** [comment]
I've just released `v2.1.0` containing this change. Thanks again!
